### ISSUE TITLE
Null values in getHashedKey function

### ISSUE THREAD

**Issue by petermein** [apiato/apiato#375]
## Expected Behavior
Null values should stay null when pass through a hashing function

## Actual Behavior
Null valued are getting hashed to an empty string

## Steps to Reproduce the Problem

  1. $entity->getHashedKey('field_name'); || $entity->getHashedKey();
  2. Where the field is nullable and value null
  3. Return ""

## Specifications
- Its a common problem when using transformers to return foreign_keys

  - Apiato Version:
2.3.1

**Comment by johannesschobel** [comment]
This is **not** an issue with apiato, but rather with the underlying HashId package. However, the author of the package clearly states, that only encoding numerical values is "allowed" ( https://github.com/ivanakimov/hashids.php#pitfalls ). Trying to encode other values (e.g., strings, null, arrays, ...) result in `""` (empty string).

I am not entirely sure if we want to change the behaviour of HashIds in Apiato, as this will then differ. Best solution, however, would be:
1) check the response for `$encodedValue === ""`
2) then return the original $value

However, in order to not break existing appliations, this **must** be a configurable value (default to have the same behaviour as now!)

-------

Regarding your specific problem:
Maybe you want to use `includes` for your `Transformer` to output related resources.

----

**Comment by johannesschobel** [comment]
@petermein @rensreinders 

ok, this has serious impact. I was just trying to think of a `configuration` solution (e.g., switching this behaviour on and off).

My "go-to" approach would be to do the following:
```
$encodedValue = $this->encoder($value);
if ($envoderValue === "") {
   return $value;
}
return $envodedValue;
```

Basically, check, if the encoder returns `""`.

**BUT** 
The encoder also returns `""` if you pass in something like this: `123qaqdfqa02342a`. If you have this described "feature" turned on, you would get the latter string as a return. However, you would not be able to tell, if the value you have received is an **encoded** value or not - because it may look like one!

The provided PR does not solve the "issue" (note the "" as i still don't think it is an issue). You may want to include related resources via `includes` in your `Transformer`. Note, that a `Transformer` also has a `null()` method that indicates an "empty" (or rather missing) relationship..

**Comment by petermein** [comment]
I understand the logic behind keeping the "". As the package of HashId's does this to keep intended behavior and type string. The problem arises only on foreign keys that are meant to be nullable. As JSON Schema dictates return values that are empty should not return anything but NULL. 
https://spacetelescope.github.io/understanding-json-schema/reference/null.html

Also includes is not an option here because includes gives you the nested resource and not the key. 
i.e.

```{
    "data": {
        "object": "DossierItemGroup",
        "id": "k0zj6aw09wlex874",
        "dossier_directory_id": "kyd9nprax5lajbv3",
        "dossier_item_group_template_id": null,
        "dossier_items": {
            "data": [
                {
                    "object": "DossierItem",
                    "id": "mn9vlbrl4w8zjoqx",
                }
            ]
        }
    },
    "meta": {
        "include": [
            "dossier_items",
           "dossier_item_group_template"
        ],
        "custom": []
    }
}```

In the above example it show that you don't always want to include the data in your application but sometimes just give the foreign key with the chance of being null. 

Our current tranformer
```
            'object' => 'DossierItemGroup',
            'id' => $entity->getHashedKey(),
            'dossier_directory_id' =>
                is_null($entity->dossier_directory_id) ? null : $entity->getHashedKey('dossier_directory_id'),
            'dossier_item_group_template_id' =>
                is_null($entity->dossier_item_group_template_id)
                    ? null
                    : $entity->getHashedKey('dossier_item_group_template_id'),

```

**Comment by johannesschobel** [comment]
I don't believe that this is a "bug" in Apiato, this may be specific to your application scenario.

As you state, this is based on "Json Schema" - in other "output formats" (e.g., a normal `DataArraySerializer`), `""` for "missing foreign keys" may be perfectly fine.

In your specific case, I would still go with the `include` approach.
You would create a `DossierItemGroupTemplateTransformer` and add respective attributes (including the `id`) to this `Transformer`. Then, in your `DossierTransformer` do the following thing:

```php
protected $defaultIncludes = [
   'dossierItemGroupTemplate'
];

public function includeDossierItemGroupTemplate(Dossier $dossier) {
   return $this->item($dossier->getDossierItemGroupTemplate()->id, new DossierItemGroupTemplateTransformer());
}
```

This will automatically (!) include the respective `Transformer` **if** a relationship exists! If you want, you can add additional logic into the respective `includeX()` method, like so:

```php
public function includeDossierItemGroupTemplate(Dossier $dossier) {
   $template = $dossier->getDossierItemGroupTemplate();

   if ($template === null) {
      return $this->null(); // this will return a "NullObject" that sets the relationship to "null"!
   }

   return $this->item($template->id, new DossierItemGroupTemplateTransformer());
```

However, if this still does not work for your specific application scenario, i would go with `extend`ing the `getHashedKey()` in your `App\Ship\Parents\Models\Model` and call the `parent` function and then extend it to your own needs.. 

Hope this helps?!
Cheers

**Comment by petermein** [comment]
In your include approach you would have to make an extra transformer to return an foreign key integer. This seems like a waist of performance. 

For now we have gone with the override the HashIdTrait with our own.
